File: src/articubot_remote/src/move_to_point_action_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "geometry_msgs/msg/point.hpp"
#include "arduinobot_msgs/action/move_to_point.hpp" // Adjust this include path
#include "moveit/move_group_interface/move_group_interface.h"

class MoveToPointActionServer : public rclcpp::Node
{
public:
  MoveToPointActionServer() : Node("move_to_point_action_server")
  {
    using namespace std::placeholders;
    this->action_server_ = rclcpp_action::create_server<arduinobot_msgs::action::MoveToPoint>(
        this,
        "move_to_point",
        std::bind(&MoveToPointActionServer::handle_goal, this, _1, _2),
        std::bind(&MoveToPointActionServer::handle_cancel, this, _1),
        std::bind(&MoveToPointActionServer::handle_accepted, this, _1));
  }

private:
  rclcpp_action::Server<arduinobot_msgs::action::MoveToPoint>::SharedPtr action_server_;

  rclcpp_action::GoalResponse handle_goal(
      const rclcpp_action::GoalUUID & uuid,
      std::shared_ptr<const arduinobot_msgs::action::MoveToPoint::Goal> goal)
  {
    RCLCPP_INFO(this->get_logger(), "Received goal request");
    (void)uuid;
    // You can reject a goal here
    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
  }

  rclcpp_action::CancelResponse handle_cancel(
      const std::shared_ptr<rclcpp_action::ServerGoalHandle<arduinobot_msgs::action::MoveToPoint>> goal_handle)
  {
    RCLCPP_INFO(this->get_logger(), "Received request to cancel goal");
    (void)goal_handle;
    return rclcpp_action::CancelResponse::ACCEPT;
  }

  void handle_accepted(const std::shared_ptr<rclcpp_action::ServerGoalHandle<arduinobot_msgs::action::MoveToPoint>> goal_handle)
  {
    using namespace std::placeholders;
    // This needs to run in a new thread to avoid blocking the executor
    std::thread{std::bind(&MoveToPointActionServer::execute, this, _1), goal_handle}.detach();
  }

  void execute(const std::shared_ptr<rclcpp_action::ServerGoalHandle<arduinobot_msgs::action::MoveToPoint>> goal_handle)
  {
    const auto goal = goal_handle->get_goal();
    auto feedback = std::make_shared<arduinobot_msgs::action::MoveToPoint::Feedback>();
    auto &distance_to_goal = feedback->distance_to_goal;
    auto result = std::make_shared<arduinobot_msgs::action::MoveToPoint::Result>();

    // Initialize MoveIt! here and move the robot arm to the goal->target_point

    // For simplicity, we assume success
    result->success = true;
    goal_handle->succeed(result);
    RCLCPP_INFO(this->get_logger(), "Goal succeeded");
  }
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<MoveToPointActionServer>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: src/articubot_remote/articubot_remote/move_to_point_action_server.py
import rclpy
from rclpy.action import ActionServer
from rclpy.node import Node

from arduinobot_msgs.action import MoveToPoint  # Update with your package name
from geometry_msgs.msg import Point
import sys

class MoveToPointActionServer(Node):

    def __init__(self):
        super().__init__('move_to_point_action_server')
        self._action_server = ActionServer(
            self,
            MoveToPoint,
            'move_to_point',
            self.execute_callback)

    async def execute_callback(self, goal_handle):
        self.get_logger().info('Executing goal...')
        feedback_msg = MoveToPoint.Feedback()
        result = MoveToPoint.Result()

        # Dummy movement logic - replace with actual MoveIt 2 integration
        target_point = goal_handle.request.target_point
        self.get_logger().info(f'Moving to point: {target_point.x}, {target_point.y}, {target_point.z}')

        # Simulate some operation
        feedback_msg.distance_to_goal = 0.0  # Update with actual distance
        self.get_logger().info('Providing feedback...')
        await goal_handle.publish_feedback(feedback_msg)

        # Assume success for simplicity
        result.success = True
        self.get_logger().info('Returning result...')
        return result

def main(args=None):
    rclpy.init(args=args)
    move_to_point_action_server = MoveToPointActionServer()

    try:
        rclpy.spin(move_to_point_action_server)
    except KeyboardInterrupt:
        pass

    rclpy.shutdown()

if __name__ == '__main__':
    main()
